### PR TITLE
XENOPS-806 Virtual Solr Spans

### PR DESCRIPTION
**Pull request by Ranecdev**

### CODE REVIEW

File: alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    public VirtualSolrSpanFactory(JSONObject jsonResponse, String path) {
+        receivingSolrInstance = path;
+
+        //QTime and Number of documents and Query found of main request
+        int numFound = -1;
+        long qTime = -1;
+        String query = "Query not found in response";
+        //Timing information if available
+        JSONObject mainTimings = null;
+
```
Comment (Ranecdev) [inline alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java:50]: Should probably have used Jackson here. Problem was that embedded jsons contained in the debug information were not uniformly/consistently named i.e: numFound vs NumFound key

Comment (vierbergenlars) [reply alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java:50]: If the structure makes it hard to use Jackson to simply have your objects with simple annotations, it's probably not worth the effort to try to use Jackson for this, and just keep your JSONObject approach.

You could do a lot of things with custom deserializers for Jackson, but I think in this case that's more complexity than what you save.

File: alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        String[] segments = in.split("/");
+        if (index == -1) {
+            return segments[segments.length - 1];
+        } else {
+            return segments[index];
+        }
+    }
+
+    private JSONObject getTimingsFromResponseJson(JSONObject jsonResponse) {
+        if (jsonResponse != null && sharded == false) {
```
Comment (vierbergenlars) [inline alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java:143]: You're calling this function from the constructor, and you're relying on some fields already being set before your function is called.

It's a bit dangerous to rely on that, as people may move code in the constructor around, which would suddenly make this function fail.

File: alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
+package eu.xenit.alfresco.instrumentation.solr;
```
[the remark this answers is not in the document]
Comment (vierbergenlars) [reply alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java:1]: I would also like to see some tests for this class.

A good start would be to capture some responses from Solr with debug information, and check if they create the expected span information.

File: alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                        long srQTime = subRequestTrackingInfo.optLong(QTIME_KEY);
+                        int srNumFound = subRequestTrackingInfo.optInt(SUB_NUMFOUND_KEY);
+                        long srElapsedTime = subRequestTrackingInfo.optLong(ELAPSEDTIME_KEY);
+                        String requestPurpose = subRequestTrackingInfo.optString(REQUEST_PURPOSE_KEY);
+
+                        String subResponse = subRequestTrackingInfo.optString("Response");
+
+                        JSONObject timings;
+                        try {
+                            // Timings section in debug information is not necessarily a correctly formatted JSONObject.
```
Comment (vierbergenlars) [inline alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java:193]: Do you have an example of this incorrectly formatted json?

> correctly formatted JSONObject

I believe you are talking about just the JSON string, before attempting to parse it, that is invalid JSON.
Referring to JSONObject (the Java class) is a bit confiusing here.

File: alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/representations/SolrRequest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package eu.xenit.alfresco.instrumentation.solr.representations;
+
+import brave.Span;
+import org.json.JSONObject;
+
+import java.util.Iterator;
+
+public class SolrRequest {
```
Comment (vierbergenlars) [inline alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/representations/SolrRequest.java:8]: Good call to make subclasses to abstract specific kinds of solr requests.

I think we can leverage abstraction a bit better by only calling methods that exist on this base class from the consumers.
That way, we don't need the specific types in `VirtualSolrSpanFactory`.

**Review by vierbergenlars — COMMENTED** [review]
A general remark: I see you have written a lot of comments in the body of methods. That is very good and helps understand sometimes not so trivial code.
However, I think it would be helpful to also have some javadoc on classes and methods that describe what that method does in general.

**Review by vierbergenlars — COMMENTED** [review]
Please add a test for non-sharded solr JSON, and I think everything is okay then

**Pull request merged**